### PR TITLE
First iteration on a better context store api

### PR DESCRIPTION
**Pull request by mar-kolya**
Separate context storage from actual fetching/putting

### CODE REVIEW

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InstrumentationContext.java [hunk 1]
```diff
@@ -5,39 +5,22 @@
...
-   * @param <V> context class
-   * @return The context instance attached to userInstance.
+   * @param <K> key class
+   * @param <C> context class
+   * @return The instance of context store for given arguments.
    */
-  public static <K, V> V get(K userInstance, Class<K> userClass, Class<V> contextClass) {
-    throw new RuntimeException("calls to this method will be rewritten");
+  public static <K, C> ContextStore<K, C> get(
+      final Class<K> keyClass, final Class<C> contextClass) {
```
Comment (realark) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InstrumentationContext.java]: Does this mean users of the api would have to cast `K` to the desired class to do the correct context lookup?

```
RunnableImpl r;

RunnableState = InstrumentationContext.get((Runnable) r, RunnableState.class);
```

Comment (mar-kolya) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InstrumentationContext.java]: No, I do not think that this statement is correct. Here is the usage example:

```
      final ContextStore<HttpURLConnection, HttpURLState> contextStore =
          InstrumentationContext.get(HttpURLConnection.class, HttpURLState.class);
      final HttpURLState state = contextStore.putIfAbsent(thiz, HttpURLState.FACTORY);
```

Comment (realark) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InstrumentationContext.java]: 👍 Oh right, we're making the two-pass lookup more explicit now. The diff threw me off.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/MapBackedProvider.java [hunk 2]
```diff
@@ -181,7 +186,7 @@ public void visitMethodInsn(
                     }
                   }
                   throw new IllegalStateException(
-                      "Incorrect Context Api Usage detected. User and context class must be class-literals. Example of correct usage: InstrumentationContext.get(runnableInstance, Runnable.class, RunnableState.class)");
+                      "Incorrect Context Api Usage detected. User and context class must be class-literals. Example of correct usage: InstrumentationContext.get(Runnable.class, RunnableContext.class)");
```
Comment (realark) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/MapBackedProvider.java]: Should be `Key and Context classes` to match the new docs.

File: dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java [hunk 3]
```diff
@@ -183,6 +190,16 @@ public static void stopSpan(
   }
 
   public static class HttpURLState {
+
+    public static Callable<HttpURLState> FACTORY =
```
Comment (realark) [inline dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java]: Should also be `final`

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/MapBackedProvider.java (hunk 2 above)
Comment (mar-kolya) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/MapBackedProvider.java]: Should be fixed now

File: dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java (hunk 3 above)
Comment (mar-kolya) [reply dd-java-agent/instrumentation/http-url-connection/src/main/java/datadog/trace/instrumentation/http_url_connection/HttpUrlConnectionInstrumentation.java]: Fixed

**Comment by mar-kolya** [comment]
@tylerbenson this seems to be a cleaner API that allows us to define operations on context storage more easily - we no longer have to rewrite every new function. It also allows us to pass context storage into functions/helpers which may come in handy.

@realark I've rebased on top of latest master and refactored some tests - could you please have another look?

File: dd-java-agent/testing/src/test/java/context/ContextTestInstrumentation.java [hunk 4]
```diff
@@ -20,59 +22,181 @@ public ContextTestInstrumentation() {
...
+    public static void methodExit() {
+      InstrumentationContext.get(KeyClass.class, Object.class);
+    }
+  }
+
+  public static class IncorrectCallContextApiUsageAdvice {
+    @Advice.OnMethodExit
+    public static void methodExit() {
+      // Our instrumentation doesn't handle variables being passed to InstrumentationContext.get,
+      // so we make sure that this actually fails instrumentation.
```
Comment (realark) [inline dd-java-agent/testing/src/test/java/context/ContextTestInstrumentation.java]: 👍

**Pull request merged**